File: src/tuner/data/sizing.py
from enum import Enum
from functools import lru_cache
from math import floor, ceil
from typing import Literal

from src.utils.static import K10, THROUGHPUT

__all__ = ['PG_DISK_SIZING']

# -----------------------------------------------------------------------------
## Note that in the list, we choose the value based on the minimum of read/write IOPS/throughput, and doing
## some averaging across value (and smally reduce the number); so the value here may not be reflected on your disk
## storage. But thinking about disk degradation after time and other factor. Thus to make use of this, it is best to
## choose based on the nearest value of minimum of read/write IOPS/throughput and round down (not up).

## These values are gained from TechPowerUp's SSD IOPS Database.
## Please note that the naming model with suffix _v1, _v2, _v3, etc. is used to differentiate the IOPS value based on
## the read/write ratio, rather than the comparison of the disk itself. Also, we don't take into the consideration of
## DRAM-availability on the SSD as we are focusing on the disk read/write after (p)SLC cache is full and not full.
## For NVME drive, when pseudo-SLC cache is full, unless you are using too small drive such as 128 GB or 256 GB, or flaw
## SSD with corrupted pSLC, or the drive is nearly full (80% full) the drive could write as around 60-80% of the
## throughput provided by the manufacturer

## For random IOPS or throughput metric, we never achieve their advertised value due to the noise, external conditions,
## but importantly the OS or driver operations, and the page-size of the I/O request. For example, the WD SN770 1 TiB
## NVME PCIe 4.0x4 has 740k IOPs read and 800k IOPs write, but measurement shows on 4K request as data disk that it
## could achieve maximum 180K IOPs and 125K - 155K for mixed read/write (50% - 99%) at 80% full capacity. Please note
## that SSD performance is degraded overtime when the drive capacity is running out due to usage, so it is best
## not to use the advertised value

class PG_DISK_SIZING(Enum):
    """
    This class contains the PostgreSQL disk sizing profile (taken from bare-metal server, not virtualized server).
    First value is the disk code
    Second value is the throughput in sequential read/write in MiB/s
    Third value is the number of IOPS in random read/write

    """
    # SATA HDDs
    HDDv1 = ('hddv1', 100, 250)     # Some old HDDs or SD cards (efficiency around 80 - 120 MiB/s)
    HDDv2 = ('hddv2', 200, 1 *K10)
    HDDv3 = ('hddv3', 260, 2500)    # Some Western Digital HDDs, or modern HDDs with small SSD/DRAM cache (220 - 290 MiB/s)

    # SAN/NAS SSDs
    SANv1 = ('sanv1', 300, 5 * K10)
    SANv2 = ('sanv2', 330, 8 * K10)
    SANv3 = ('sanv3', 370, 12 * K10)
    SANv4 = ('sanv4', 400, 16 * K10)

    # SATA SSDs (Local)
    SSDv1 = ('ssdv1', 450, 20 * K10)
    SSDv2 = ('ssdv2', 500, 30 * K10)
    SSDv3 = ('ssdv3', 533, 40 * K10)
    SSDv4 = ('ssdv4', 566, 50 * K10)
    SSDv5 = ('ssdv5', 600, 60 * K10)

    # Remote NVMe SSD (Usually the NVMe Box)
    NVMeBOXv1 = ('nvmeboxv1', 800, 80 * K10)
    NVMeBOXv2 = ('nvmeboxv2', 1000, 100 * K10)
    NVMeBOXv3 = ('nvmeboxv3', 1400, 120 * K10)
    NVMeBOXv4 = ('nvmeboxv4', 1700, 140 * K10)

    # We don't make custom value of NVME SSD such as PCIe 3.0/4.0 x8 or 3.0/4.0 x16, as it's extremely rare to nothing.
    # NVMe PCIe Gen 3 SSDs
    NVMePCIev3x4v1 = ('nvmepciev3x4v1', 2000, 150 * K10)
    NVMePCIev3x4v2 = ('nvmepciev3x4v2', 2500, 200 * K10)
    NVMePCIev3x4v3 = ('nvmepciev3x4v3', 3000, 250 * K10)
    NVMePCIev3x4v4 = ('nvmepciev3x4v4', 3500, 300 * K10)
    NVMePCIev3x4v5 = ('nvmepciev3x4v5', 4000, 350 * K10)
    NVMePCIev3x4v6 = ('nvmepciev3x4v6', 4500, 400 * K10)
    # NVMe PCIe Gen 4 SSDs
    NVMePCIev4x4v1 = ('nvmepciev4x4v1', 4500, 300 * K10)
    NVMePCIev4x4v2 = ('nvmepciev4x4v2', 5000, 375 * K10)
    NVMePCIev4x4v3 = ('nvmepciev4x4v3', 5500, 450 * K10)
    NVMePCIev4x4v4 = ('nvmepciev4x4v4', 6000, 525 * K10)
    NVMePCIev4x4v5 = ('nvmepciev4x4v5', 6500, 600 * K10)
    NVMePCIev4x4v6 = ('nvmepciev4x4v6', 7000, 700 * K10)
    # NVMe PCIe Gen 5 SSDs
    NVMePCIev5x4v1 = ('nvmepciev5x4v1', 7000, 750 * K10)
    NVMePCIev5x4v2 = ('nvmepciev5x4v2', 8500, 850 * K10)
    NVMePCIev5x4v3 = ('nvmepciev5x4v3', 9500, 950 * K10)
    NVMePCIev5x4v4 = ('nvmepciev5x4v4', 11000, 1100 * K10)
    NVMePCIev5x4v5 = ('nvmepciev5x4v5', 12500, 1250 * K10)
    NVMePCIev5x4v6 = ('nvmepciev5x4v6', 14000, 1400 * K10)

    @staticmethod
    @lru_cache(maxsize=1)
    def _disk_type_list_v1() -> list[str]:
        return ['hdd', 'san', 'ssd', 'nvmebox', 'nvmepciev3', 'nvmepciev4', 'nvmepciev5']

    @staticmethod
    @lru_cache(maxsize=1)
    def _disk_type_list_v2() -> list[str]:
        return ['hdd', 'san', 'ssd', 'nvmebox', 'nvmepciev3', 'nvmepciev4', 'nvmepciev5', 'nvmepcie', 'nvme']

    def disk_code(self) -> str:
        return self.value[0]

    def throughput(self) -> int:
        return self.value[1]

    def iops(self) -> int:
        return self.value[2]

    def _check_disk_type(self, disk_type: str) -> bool:
        disk_type = disk_type.lower()
        if disk_type not in PG_DISK_SIZING._disk_type_list_v2():
            raise ValueError(f'Disk type {disk_type} is not available')
        return self.disk_code().startswith(disk_type)

    @staticmethod
    @lru_cache(maxsize=32)
    def _list(disk_type: str | None, performance_type: str | None = None) -> list['PG_DISK_SIZING']:
        result = [disk for disk in PG_DISK_SIZING if disk_type is None or disk._check_disk_type(disk_type)]
        if performance_type is not None:
            fn = lambda x: (x.throughput(), x.iops()) if performance_type == THROUGHPUT else (x.iops(), x.throughput())
            result.sort(key=fn, reverse=False)
        return result

    @staticmethod
    def _find_midpoints(disks: list['PG_DISK_SIZING'], performance_type: str) -> int | float:
        midpoint, remainder = divmod(len(disks), 2)
        # For even number of disks, we choose between two disks
        if remainder == 0:
            tmp_disk01 = disks[midpoint - 1]
            tmp_disk02 = disks[midpoint]
            if performance_type == THROUGHPUT:
                spec = (tmp_disk01.throughput() + tmp_disk02.throughput()) / 2
            else:
                spec = (tmp_disk01.iops() + tmp_disk02.iops()) / 2
        else:
            if performance_type == THROUGHPUT:
                spec = disks[midpoint + remainder].throughput()
            else:
                spec = disks[midpoint + remainder].iops()
        return spec

    # Maximum number of cache entry (if just int) is 2 * (num_type_of_disk * 2) ^ 2
    # But we never reach that. The current codebase (Jan 28th 2025 said we only have 13 entries only).
    # So for future-proof, we would use 32 as the maximum cache entry
    @staticmethod
    @lru_cache(maxsize=32)
    def _get_bound(performance_type: str, disk_01: 'PG_DISK_SIZING', disk_02: 'PG_DISK_SIZING') -> tuple[int, int]:
        """
        This function try to get the upper and lower bound for comparison. Designed to leverage the lru_cache
        caching mechanism.

        Arguments:
        ---------

        performance_type: str
            The performance type, either THROUGHPUT or RANDOM_IOPS

        disk_01: PG_DISK_SIZING | int | float
            The first disk or the performance value. If the input is numeric, we used it as the definite bound with
            no resizing.

        disk_02: PG_DISK_SIZING | int | float
            The second disk or the performance value. If the input is numeric, we used it as the definite bound with
            no resizing.

        Returns:
        -------

        tuple[int, int]
            The lower and upper bound for comparison
        """
        _disk_table = PG_DISK_SIZING._list(disk_type=None, performance_type=performance_type)
        if isinstance(disk_01, PG_DISK_SIZING):
            disk01_index = _disk_table.index(disk_01)
            prev_disk01 = _disk_table[max(0, disk01_index - 1)]
            # Deal when no weaker disk
            if disk_01 == prev_disk01:
                lower_bound = 0
            elif performance_type == THROUGHPUT:
                lower_bound = (disk_01.throughput() + prev_disk01.throughput()) / 2
            else:
                lower_bound = (disk_01.iops() + prev_disk01.iops()) / 2
        else:
            lower_bound = disk_01

        if isinstance(disk_02, PG_DISK_SIZING):
            disk02_index = _disk_table.index(disk_02)
            latt_disk02 = _disk_table[min(len(_disk_table) - 1, disk02_index + 1)]
            # Deal when no stronger disk
            if disk_02 == latt_disk02:
                # Any positive number larger than 1
                upper_bound = 2 * (disk_02.throughput() if performance_type == THROUGHPUT else disk_02.iops())
            elif performance_type == THROUGHPUT:
                upper_bound = (disk_02.throughput() + latt_disk02.throughput()) / 2
            else:
                upper_bound = (disk_02.iops() + latt_disk02.iops()) / 2
        else:
            upper_bound = disk_02

        # Swap if the lower bound is greater than the upper bound
        if upper_bound < lower_bound:
            lower_bound, upper_bound = upper_bound, lower_bound

        return floor(lower_bound), ceil(upper_bound)

    @staticmethod
    def match_between(performance: int, performance_type: str, disk_01: 'PG_DISK_SIZING',
                      disk_02: 'PG_DISK_SIZING') -> bool:
        # Fill the gap when the expected disk is strong (probably when we allowed higher disk performance)
        # but the disk here is not available
        _disk_table = PG_DISK_SIZING._list(disk_type=None, performance_type=performance_type)
        if performance_type == THROUGHPUT and performance >= _disk_table[-1].throughput():
            return True
        elif performance_type != THROUGHPUT and performance >= _disk_table[-1].iops():
            return True

        lower_bound, upper_bound = PG_DISK_SIZING._get_bound(performance_type, disk_01, disk_02)
        return lower_bound <= performance < upper_bound

    @staticmethod
    def match_disk_series(performance: int, performance_type: str, disk_type: str,
                          interval: Literal['all', 'weak', 'strong'] = 'all') -> bool:
        disks = PG_DISK_SIZING._list(disk_type, performance_type)
        if not disks:
            raise ValueError(f'No disk type found when matching {disk_type}')
        if interval == 'all':
            return PG_DISK_SIZING.match_between(performance, performance_type, disks[0], disks[-1])

        middle_specification = PG_DISK_SIZING._find_midpoints(disks, performance_type=performance_type)
        pairs = (disks[0], middle_specification) if interval == 'weak' else (middle_specification, disks[-1])
        return PG_DISK_SIZING.match_between(performance, performance_type, *pairs)

    @staticmethod
    def match_one_disk(performance: int, performance_type: str, disk: 'PG_DISK_SIZING') -> bool:
        return PG_DISK_SIZING.match_between(performance, performance_type, disk_01=disk, disk_02=disk)

    @staticmethod
    def match_disk_series_in_range(performance: int, performance_type: str, disk01_type: str, disk02_type: str) -> bool:
        # Check if we are in between the transition between two disks
        if disk01_type == disk02_type:
            return PG_DISK_SIZING.match_disk_series(performance, performance_type, disk_type=disk01_type)

        disk01s = PG_DISK_SIZING._list(disk01_type, performance_type)
        disk02s = PG_DISK_SIZING._list(disk02_type, performance_type)

        if not disk01s or not disk02s:
            raise ValueError(f'No disk type found when matching {disk01_type} and {disk02_type}')

        disk_collection = [disk01s[0], disk01s[-1], disk02s[0], disk02s[-1]]
        fn = lambda x: (x.throughput(), x.iops()) if performance_type == THROUGHPUT else (x.iops(), x.throughput())
        disk_collection.sort(key=fn, reverse=False)
        return PG_DISK_SIZING.match_between(performance, performance_type, disk_01=disk_collection[0],
                                            disk_02=disk_collection[-1])

File: js/tuner/data/sizing.js
// --------------------------------------------------------------------------
import { K10, THROUGHPUT } from '../../static/vars.js';

// -------------------------------------------------------------
// ENUM choices
const SIZE_PROFILE = ['mini', 'medium', 'large', 'mall', 'bigt'];
const _ascending_specs = {
    size: SIZE_PROFILE,
    vcpu_min: [1, 2, 6, 12, 32],
    vcpu_max: [4, 8, 16, 48, 128],
    ram_gib_min: [2, 8, 24, 48, 128],
    ram_gib_max: [16, 32, 64, 192, 512],
    storage_gib_max: [50, 300, 1024, 5120, 32768],
    network_mbps_max: [500, 1000, 5000, 12500, 30000],
};
function _str_to_num(value) {
    return _ascending_specs.size.indexOf(value);
}

// --------------------------------------------------------------
// PG_SIZING: Represents a PostgreSQL sizing profile
class PG_SIZING {
    constructor(value) {
        this.value = value; // one of 'mini', 'medium', etc.
    }

    num() {
        return _str_to_num(this.value);
    }

    lt(other) {
        return this.num() < other.num();
    }

    eq(other) {
        return this.num() === other.num();
    }
}

// Define the PG_SIZING enum members
PG_SIZING.MINI = new PG_SIZING(SIZE_PROFILE[0]);
PG_SIZING.MEDIUM = new PG_SIZING(SIZE_PROFILE[1]);
PG_SIZING.LARGE = new PG_SIZING(SIZE_PROFILE[2]);
PG_SIZING.MALL = new PG_SIZING(SIZE_PROFILE[3]);
PG_SIZING.BIGT = new PG_SIZING(SIZE_PROFILE[4]);

// ----------------------------------------------------------------
// PG_DISK_SIZING: Represents a PostgreSQL disk sizing profile
class PG_DISK_SIZING {
    constructor(code, throughput, iops) {
        this._code = code;
        this._throughput = throughput;
        this._iops = iops;
    }

    diskCode() {
        return this._code;
    }

    throughput() {
        return this._throughput;
    }

    iops() {
        return this._iops;
    }

    _checkDiskType(diskType) {
        const dt = diskType.toLowerCase();
        if (!PG_DISK_SIZING._diskTypeListV2().includes(dt)) {
            throw new Error(`Disk type ${dt} is not available`);
        }
        return this._code.startsWith(dt);
    }

    // Static helper methods and caching
    static _diskTypeListV1() {
        return ['hdd', 'san', 'ssd', 'nvmebox', 'nvmepciev3', 'nvmepciev4', 'nvmepciev5'];
    }

    static _diskTypeListV2() {
        return ['hdd', 'san', 'ssd', 'nvmebox', 'nvmepciev3', 'nvmepciev4', 'nvmepciev5', 'nvmepcie', 'nvme'];
    }

    static _all() {
        return PG_DISK_SIZING.ALL;
    }

    static _list(diskType = null, performanceType = null) {
        let result = PG_DISK_SIZING._all().filter(disk => {
            return diskType === null || disk._checkDiskType(diskType);
        });
        if (performanceType !== null) {
            const keyFn = performanceType === THROUGHPUT ?
                (d) => [d.throughput(), d.iops()] :
                (d) => [d.iops(), d.throughput()];
            result.sort((a, b) => {
                const ka = keyFn(a), kb = keyFn(b);
                return ka[0] - kb[0] || ka[1] - kb[1];
            });
        }
        return result;
    }

    static _findMidpoints(disks, performanceType) {
        const len = disks.length;
        const midpoint = Math.floor(len / 2);
        if (len % 2 === 0) {
            const disk1 = disks[midpoint - 1];
            const disk2 = disks[midpoint];
            return performanceType === THROUGHPUT ?
                (disk1.throughput() + disk2.throughput()) / 2 :
                (disk1.iops() + disk2.iops()) / 2;
        } else {
            const disk = disks[midpoint];
            return performanceType === THROUGHPUT ? disk.throughput() : disk.iops();
        }
    }

    static _getBound(performanceType, disk01, disk02) {
        const diskTable = PG_DISK_SIZING._list(null, performanceType);
        let lowerBound;
        if (disk01 instanceof PG_DISK_SIZING) {
            const idx = diskTable.indexOf(disk01);
            const prev = idx > 0 ? diskTable[idx - 1] : disk01;
            lowerBound = disk01 === prev ? 0 :
                (performanceType === THROUGHPUT ?
                    (disk01.throughput() + prev.throughput()) / 2 :
                    (disk01.iops() + prev.iops()) / 2);
        } else {
            lowerBound = disk01;
        }

        let upperBound;
        if (disk02 instanceof PG_DISK_SIZING) {
            const idx = diskTable.indexOf(disk02);
            const next = idx < diskTable.length - 1 ? diskTable[idx + 1] : disk02;
            upperBound = disk02 === next ?
                2 * (performanceType === THROUGHPUT ? disk02.throughput() : disk02.iops()) :
                (performanceType === THROUGHPUT ?
                    (disk02.throughput() + next.throughput()) / 2 :
                    (disk02.iops() + next.iops()) / 2);
        } else {
            upperBound = disk02;
        }

        if (upperBound < lowerBound) {
            [lowerBound, upperBound] = [upperBound, lowerBound];
        }

        return [Math.floor(lowerBound), Math.ceil(upperBound)];
    }

    static matchBetween(performance, performanceType, disk01, disk02) {
        const diskTable = PG_DISK_SIZING._list(null, performanceType);
        const lastDisk = diskTable[diskTable.length - 1];
        if (performanceType === THROUGHPUT && performance >= lastDisk.throughput()) {
            return true;
        } else if (performanceType !== THROUGHPUT && performance >= lastDisk.iops()) {
            return true;
        }
        const [lowerBound, upperBound] = PG_DISK_SIZING._getBound(performanceType, disk01, disk02);
        return performance >= lowerBound && performance < upperBound;
    }

    static matchDiskSeries(performance, performanceType, diskType, interval = 'all') {
        const disks = PG_DISK_SIZING._list(diskType, performanceType);
        if (!disks.length) {
            throw new Error(`No disk type found when matching ${diskType}`);
        }
        if (interval === 'all') {
            return PG_DISK_SIZING.matchBetween(performance, performanceType, disks[0], disks[disks.length - 1]);
        }
        if (interval === 'weak') {
            return PG_DISK_SIZING.matchBetween(performance, performanceType, disks[0], disks[Math.floor(disks.length / 2)]);
        } else { // 'strong'
            return PG_DISK_SIZING.matchBetween(performance, performanceType, disks[Math.floor(disks.length / 2)], disks[disks.length - 1]);
        }
    }

    static matchOneDisk(performance, performanceType, disk) {
        return PG_DISK_SIZING.matchBetween(performance, performanceType, disk, disk);
    }

    static matchDiskSeriesInRange(performance, performanceType, disk01Type, disk02Type) {
        if (disk01Type === disk02Type) {
            return PG_DISK_SIZING.matchDiskSeries(performance, performanceType, disk01Type);
        }
        const disk01s = PG_DISK_SIZING._list(disk01Type, performanceType);
        const disk02s = PG_DISK_SIZING._list(disk02Type, performanceType);
        if (!disk01s.length || !disk02s.length) {
            throw new Error(`No disk type found when matching ${disk01Type} and ${disk02Type}`);
        }
        const diskCollection = [
            disk01s[0],
            disk01s[disk01s.length - 1],
            disk02s[0],
            disk02s[disk02s.length - 1]
        ];
        const sortFn = performanceType === THROUGHPUT ?
            (a, b) => a.throughput() - b.throughput() || a.iops() - b.iops() :
            (a, b) => a.iops() - b.iops() || a.throughput() - b.throughput();
        diskCollection.sort(sortFn);
        return PG_DISK_SIZING.matchBetween(performance, performanceType, diskCollection[0], diskCollection[diskCollection.length - 1]);
    }
}

// Define the PG_DISK_SIZING enum members
// SATA HDDs
PG_DISK_SIZING.HDDv1 = new PG_DISK_SIZING('hddv1', 100, 250);
PG_DISK_SIZING.HDDv2 = new PG_DISK_SIZING('hddv2', 200, K10);
PG_DISK_SIZING.HDDv3 = new PG_DISK_SIZING('hddv3', 260, 2500);

// SAN/NAS SSDs
PG_DISK_SIZING.SANv1 = new PG_DISK_SIZING('sanv1', 300, 5 * K10);
PG_DISK_SIZING.SANv2 = new PG_DISK_SIZING('sanv2', 330, 8 * K10);
PG_DISK_SIZING.SANv3 = new PG_DISK_SIZING('sanv3', 370, 12 * K10);
PG_DISK_SIZING.SANv4 = new PG_DISK_SIZING('sanv4', 400, 16 * K10);

// SATA SSDs (Local)
PG_DISK_SIZING.SSDv1 = new PG_DISK_SIZING('ssdv1', 450, 20 * K10);
PG_DISK_SIZING.SSDv2 = new PG_DISK_SIZING('ssdv2', 500, 30 * K10);
PG_DISK_SIZING.SSDv3 = new PG_DISK_SIZING('ssdv3', 533, 40 * K10);
PG_DISK_SIZING.SSDv4 = new PG_DISK_SIZING('ssdv4', 566, 50 * K10);
PG_DISK_SIZING.SSDv5 = new PG_DISK_SIZING('ssdv5', 600, 60 * K10);

// Remote NVMe SSD (Usually the NVMe Box)
PG_DISK_SIZING.NVMeBOXv1 = new PG_DISK_SIZING('nvmeboxv1', 800, 80 * K10);
PG_DISK_SIZING.NVMeBOXv2 = new PG_DISK_SIZING('nvmeboxv2', 1000, 100 * K10);
PG_DISK_SIZING.NVMeBOXv3 = new PG_DISK_SIZING('nvmeboxv3', 1400, 120 * K10);
PG_DISK_SIZING.NVMeBOXv4 = new PG_DISK_SIZING('nvmeboxv4', 1700, 140 * K10);

// NVMe PCIe Gen 3 SSDs
PG_DISK_SIZING.NVMePCIev3x4v1 = new PG_DISK_SIZING('nvmepciev3x4v1', 2000, 150 * K10);
PG_DISK_SIZING.NVMePCIev3x4v2 = new PG_DISK_SIZING('nvmepciev3x4v2', 2500, 200 * K10);
PG_DISK_SIZING.NVMePCIev3x4v3 = new PG_DISK_SIZING('nvmepciev3x4v3', 3000, 250 * K10);
PG_DISK_SIZING.NVMePCIev3x4v4 = new PG_DISK_SIZING('nvmepciev3x4v4', 3500, 300 * K10);
PG_DISK_SIZING.NVMePCIev3x4v5 = new PG_DISK_SIZING('nvmepciev3x4v5', 4000, 350 * K10);
PG_DISK_SIZING.NVMePCIev3x4v6 = new PG_DISK_SIZING('nvmepciev3x4v6', 4500, 400 * K10);

// NVMe PCIe Gen 4 SSDs
PG_DISK_SIZING.NVMePCIev4x4v1 = new PG_DISK_SIZING('nvmepciev4x4v1', 4500, 300 * K10);
PG_DISK_SIZING.NVMePCIev4x4v2 = new PG_DISK_SIZING('nvmepciev4x4v2', 5000, 375 * K10);
PG_DISK_SIZING.NVMePCIev4x4v3 = new PG_DISK_SIZING('nvmepciev4x4v3', 5500, 450 * K10);
PG_DISK_SIZING.NVMePCIev4x4v4 = new PG_DISK_SIZING('nvmepciev4x4v4', 6000, 525 * K10);
PG_DISK_SIZING.NVMePCIev4x4v5 = new PG_DISK_SIZING('nvmepciev4x4v5', 6500, 600 * K10);
PG_DISK_SIZING.NVMePCIev4x4v6 = new PG_DISK_SIZING('nvmepciev4x4v6', 7000, 700 * K10);

// NVMe PCIe Gen 5 SSDs
PG_DISK_SIZING.NVMePCIev5x4v1 = new PG_DISK_SIZING('nvmepciev5x4v1', 7000, 750 * K10);
PG_DISK_SIZING.NVMePCIev5x4v2 = new PG_DISK_SIZING('nvmepciev5x4v2', 8500, 850 * K10);
PG_DISK_SIZING.NVMePCIev5x4v3 = new PG_DISK_SIZING('nvmepciev5x4v3', 9500, 950 * K10);
PG_DISK_SIZING.NVMePCIev5x4v4 = new PG_DISK_SIZING('nvmepciev5x4v4', 11000, 1100 * K10);
PG_DISK_SIZING.NVMePCIev5x4v5 = new PG_DISK_SIZING('nvmepciev5x4v5', 12500, 1250 * K10);
PG_DISK_SIZING.NVMePCIev5x4v6 = new PG_DISK_SIZING('nvmepciev5x4v6', 14000, 1400 * K10);

// Populate ALL list
PG_DISK_SIZING.ALL = [
    PG_DISK_SIZING.HDDv1, PG_DISK_SIZING.HDDv2, PG_DISK_SIZING.HDDv3,
    PG_DISK_SIZING.SANv1, PG_DISK_SIZING.SANv2, PG_DISK_SIZING.SANv3, PG_DISK_SIZING.SANv4,
    PG_DISK_SIZING.SSDv1, PG_DISK_SIZING.SSDv2, PG_DISK_SIZING.SSDv3, PG_DISK_SIZING.SSDv4, PG_DISK_SIZING.SSDv5,
    PG_DISK_SIZING.NVMeBOXv1, PG_DISK_SIZING.NVMeBOXv2, PG_DISK_SIZING.NVMeBOXv3, PG_DISK_SIZING.NVMeBOXv4,
    PG_DISK_SIZING.NVMePCIev3x4v1, PG_DISK_SIZING.NVMePCIev3x4v2, PG_DISK_SIZING.NVMePCIev3x4v3,
    PG_DISK_SIZING.NVMePCIev3x4v4, PG_DISK_SIZING.NVMePCIev3x4v5, PG_DISK_SIZING.NVMePCIev3x4v6,
    PG_DISK_SIZING.NVMePCIev4x4v1, PG_DISK_SIZING.NVMePCIev4x4v2, PG_DISK_SIZING.NVMePCIev4x4v3,
    PG_DISK_SIZING.NVMePCIev4x4v4, PG_DISK_SIZING.NVMePCIev4x4v5, PG_DISK_SIZING.NVMePCIev4x4v6,
    PG_DISK_SIZING.NVMePCIev5x4v1, PG_DISK_SIZING.NVMePCIev5x4v2, PG_DISK_SIZING.NVMePCIev5x4v3,
    PG_DISK_SIZING.NVMePCIev5x4v4, PG_DISK_SIZING.NVMePCIev5x4v5, PG_DISK_SIZING.NVMePCIev5x4v6,
];

export { PG_DISK_SIZING, PG_SIZING, SIZE_PROFILE };
